compiler/passes/src/type_checking/check_expressions.rs

```diff
@@ -101,11 +101,36 @@ impl<'a> ExpressionVisitor<'a> for TypeChecker<'a> {
                 match *access.inner {
                     // If the access expression is of the form `self.<name>`, then check the <name> is valid.
                     Expression::Identifier(identifier) if identifier.name == sym::SelfLower => match access.name.name {
-                        sym::caller => return Some(Type::Address),
+                        sym::caller => {
+                            // Check that the operation is not invoked in a `finalize` block.
+                            if self.is_finalize {
+                                self.handler.emit_err(TypeCheckerError::invalid_operation_inside_finalize(
+                                    "self.caller",
+                                    access.name.span(),
+                                ))
+                            }
+                            return Some(Type::Address);
+                        }
```

Comment:
Is this a new requirement for `self.caller` to be only usable in finalize blocks?